### ISSUE TITLE
kNN visualization (interpretation)

### ISSUE THREAD

**Issue by dekaufman** [yzhao062/pyod#77]
The visualization produced by `knn_example.py` for the "Test Set Prediction" shows two false positives, i.e., 12 outlier findings instead of 10 as in the "Test Set Ground Truth" chart.  Isn't this somewhat inconsistent with the result printed to console that ROC_AUC = 1?

If so, I think the inconsistency arises because the predicted labels for the chart are based on `y_test_pred = clf.predict(X_test)`.  I think that means that the test labels are being predicted from comparison the distance threshold `clf.threshold_` obtained in fitting `clf` to the *training* data.  In contrast, the ROC_AUC value is based on a fixed contamination rate (10%).

It would only make sense to use `clf.threshold_` for this purpose if the kNN distance for any point x_i in the test set were being computed over the distances from x_i to each of the 200 training points, not the distances to the other 99 test points.  But then, the ROC_AUC curve ought to be based on those same labels, and it isn't, is it?  I think it's currently being generated from a set of labels that re-applies the 10% contamination assumption, ignoring `clf.threshold_`.

(I can't quite follow whether the kNN distances for the points in the test set are being computed vs. the training set or vs. the other points in the test set.  Can you clarify this for me?  I have to guess that it's the former; if it's the latter, then it would seem really weird to be applying `clf.threshold_` from a training set of a different size.)

Is it even appropriate to apply the kNN model from training data directly to the test set?  I would have thought this use of kNN is intended to be used on an entire data set all by itself.  Although one could perhaps study a training set to make a reasonable judgment about appropriate values of `k` and contamination rate.

Thanks!

**Comment by yzhao062** [comment]
Hi David, 
Thanks for the note. Your concern is totally legit.

The roc score is measured by (y_test, t_test_scores), so it regards the test points as a group but ignores the existence of the training samples. For instance, 

```
y_test (ground truth): [0,1,0,1,0]
y_test_scores (predicted by pyod models): [0.1, 0.4, 0.2, 0.5, 0.3]
```

then the roc score is 1 as the rank is reserved.

However, the conversion from raw scores to binary labels is different. This step is inferred by comparing the test scores with the absolute threshold, `clf_threshold_`, generated in `fit(X_train)`. `clf_threshold_` is inferred from contamination parameter. 
![image](https://user-images.githubusercontent.com/15079146/56433417-cd495200-629e-11e9-957f-61193ca1c2ce.png)

![image](https://user-images.githubusercontent.com/15079146/56433026-8c047280-629d-11e9-894d-4e6a498470e1.png)

Therefore, the problem is roc evaluation on `y_test` is independent to `contamination` parameter and the existence of training samples. The relationship is only weakly inferred.

I fully understand this behavior sounds shaky but there is a limitation in unsupervised type inference, whether should we actually `predict` or combine the train and test to do another training. Another point is people do not usually use binary labels in outlier detection but raw scores, so the confusion somehow is alleviated. 

I will think about what is the better way to prevent the confusion in demonstrating the library via examples. Thanks for the note and hope this at least make it a bit clearer.
